### PR TITLE
bf: S3C-4055 cloudserver metrics computation correction

### PR DESCRIPTION
**Pull request by naren-rajendran**
metrics like incomingBytes, sizeDelta were not calculated correctly for actions like getObject and it is corrected now.

### CODE REVIEW

**Comment by rahulreddy** [comment]
@naren-scality Can you please change the base branch to `stabilization/7.9.0`? It needs to be picked up for 7.9.0

File: tests/utapi/utilities.js [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+    testEvents.forEach(event => {
+        it(`should compute and push metrics for ${event.action}`, () => {
+            const eventPushed = pushMetric(event.action, log, event.metrics);
+            assert(eventPushed);
+            const compareObj = computeMetrics(event.action, event.metrics);
+            assert(compareObj);
+            Object.keys(compareObj).forEach(key => {
+                assert.strictEqual(eventPushed[key], compareObj[key]);
+            });
```
Comment (miniscruff) [inline tests/utapi/utilities.js]: I think this can be assert.deepStrictEqual instead.

File: tests/utapi/utilities.js [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    const incomingBytes = action === 'getObject' ? 0 : newByteLength;
+    const sizeDelta = oldByteLength ? newByteLength - oldByteLength :
+        incomingBytes;
+    return {
+        objectDelta: isDelete ? -numberOfObjects : numberOfObjects,
+        sizeDelta: isDelete ? -byteLength : sizeDelta,
+        incomingBytes,
+        outgoingBytes: action === 'getObject' ? newByteLength : 0,
+    };
+}
```
Comment (miniscruff) [inline tests/utapi/utilities.js]: Duplicating the logic into the test as the code itself isn't the best method of writing tests. For one it ties our tests to the internal logic which we should not do, we just want to test the observable result (eg the output). Additionally, this just proves that what comes in matches what comes out and not that what comes out is correct. For example, if you wrote the test by copy pasting the broken code, the test would still pass even though it was wrong.

Instead, create `expected` objects to compare to in your test array.

Basically it would be the same as doing: 
```js
const jsonString = '{"a": 5}';
assert.strictEqual(JSON.stringify(JSON.parse(jsonString)), jsonString);
```

**Review by miniscruff — CHANGES_REQUESTED** [review]

File: tests/utapi/utilities.js (hunk 2 above)
Comment (naren-rajendran) [reply tests/utapi/utilities.js]: I am still working on this. This is not quite up for review yet, will convert this to draft. I have the correction but did not push it yet... too soon @miniscruff :)

**Comment by naren-rajendran** [comment]
@bert-e reset

**Review by anurag4DSB — COMMENTED** [review]
Do we not have any tests yet?

**Comment by naren-rajendran** [comment]
> Do we not have any tests yet?

@anurag4DSB On cloudserver, no, we do not! We have integration tests for utapi v1 but not for v2.

**Review by anurag4DSB — APPROVED** [review]

**Review by tmacro — APPROVED** [review]

**Comment by naren-rajendran** [comment]
@bert-e reset

File: tests/utapi/utilities.js [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+        sinon.restore();
+    });
+
+    testEvents.forEach(event => {
+        it(`should compute and push metrics for ${event.action}`, () => {
+            const eventPushed = pushMetric(event.action, log, event.metrics);
+            assert(eventPushed);
+            Object.keys(event.expected).forEach(key => {
+                assert.deepStrictEqual(eventPushed[key], event.expected[key]);
+            });
```
Comment (miniscruff) [inline tests/utapi/utilities.js]: I think just one assertion will work here.
```suggestion
            assert.deepStrictEqual(eventPushed, event.expected);
```

File: tests/utilities/mock/Utapi.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
+const express = require('express');
```
Comment (miniscruff) [inline tests/utilities/mock/Utapi.js:1]: This feels like a lot of code for a mock. Feels like we would need to add tests here as it has a fair bit of logic.

**Comment by naren-rajendran** [comment]
@bert-e reset

File: tests/utapi/awsNodeSdk.js [hunk 5]
```diff
@@ -0,0 +1,375 @@
+const async = require('async');
```
Comment (miniscruff) [inline tests/utapi/awsNodeSdk.js:1]: I feel like this isn't wrong but a bit much. Both with the mock UTAPI server with express and recreating a bunch of events to send to UTAPI. Part of my concern here is too much testing as well as potentially duplicate testing.
It seems to me our unit testing and functional testing overlap here as well as what would probably be included in UTAPI v2. 

I don't really want to block the PR to come up with some sort of test plan or Venn Diagram of what is tested where but I don't want to blindly review all this code.

**Review by miniscruff — CHANGES_REQUESTED** [review]

File: tests/utapi/utilities.js (hunk 3 above)
Comment (naren-rajendran) [reply tests/utapi/utilities.js]: Not all properties of the published event are compared, only metrics.

**Review by miniscruff — APPROVED** [review]

**Comment by naren-rajendran** [comment]
@bert-e approve

**Comment by naren-rajendran** [comment]
@bert-e reset

**Comment by naren-rajendran** [comment]
@bert-e reset

**Comment by naren-rajendran** [comment]
@bert-e reset

**Comment by naren-rajendran** [comment]
@bert-e help

**Comment by naren-rajendran** [comment]
@bert-e status

**Comment by naren-rajendran** [comment]
@bert-e reset

**Comment by naren-rajendran** [comment]
@bert-e help

**Comment by naren-rajendran** [comment]
@bert-e reset

**Comment by naren-rajendran** [comment]
@bert-e reset

**Comment by naren-rajendran** [comment]
@bert-e force_reset

**Comment by rahulreddy** [comment]
I am going to break 8.1 and 8.2 branches by merging this to allow 7.9 RC4 to be built. @naren-scality  will follow up with a fix for 8.1 and 8.2 branches.
